tests-functional/tests/test_wakuext_community_token_permissions.py

```diff
@@ -0,0 +1,286 @@
+            f"cast send {self.controller_address} 'generateTokens(address,uint256)' "
+            f"{member_address} {token_amount} --rpc-url http://anvil:8545 "
+            f"--private-key {user_1.private_key}"
+        )
+        result = foundry_client.container.exec_run(generate_cmd)
+        logger.debug(f"Generate tokens result for member: exit_code={result.exit_code}, output={result.output.decode()}")
+
+        logger.debug(f"Funded {member_address} with 10 SNT tokens at contract {self.snt_address}")
+
+    def create_token_gated_community(
+        self, owner_backend, permission_type=CommunityTokenPermissionType.BECOME_MEMBER, membership=CommunityPermissionsAccess.AUTO_ACCEPT
+    ):
+        """Helper to create a community with token permissions"""
+        # Create basic community
+        community_resp = owner_backend.wakuext_service.create_community(
+            name=fake.community_name(),
+            description=fake.community_description(),
+            membership=membership,
+        )
+        community_id = community_resp.get("communities", [{}])[0].get("id")
+
+        # Add token permission
+        token_criteria = [
+            {
+                "type": CommunityTokenType.ERC20,
+                "contract_addresses": {31337: "0x1234567890123456789012345678901234567890"},
+                "symbol": "TEST",
+                "amountInWei": "1",  # 1 wei required
+                "decimals": 18,
+            }
+        ]
+
+        permission_resp = owner_backend.wakuext_service.create_community_token_permission(
+            community_id=community_id, permission_type=permission_type, token_criteria=token_criteria
+        )
+
+        return community_id, permission_resp
+
+    @pytest.mark.skip(reason="Pending on issue https://github.com/status-im/status-go/issues/7114")
+    def test_token_gated_community_membership_no_valid_tokens(self):
+        """Test that users must hold required tokens to join community"""
+        # Owner creates token-gated community
+        community_id, _ = self.create_token_gated_community(
+            self.owner, CommunityTokenPermissionType.BECOME_MEMBER, CommunityPermissionsAccess.MANUAL_ACCEPT
+        )
+
+        # Fetch community as member
+        self.fetch_community(self.member, community_id)
+
+        # Member tries to join without tokens - should fail at request time
+        join_req = self.member.wakuext_service.request_to_join_community(community_id, self.fake_address)
+        requests = join_req.get("requestsToJoinCommunity", [])
+        if requests:
+            # If request was created, check that it gets declined
+            req_id = requests[0].get("id")
+            # Check that request is declined due to insufficient permissions
+            declined_reqs = retry_call(self.owner.wakuext_service.declined_requests_to_join_for_community, community_id)
+            assert len(declined_reqs) == 1
+            assert declined_reqs[0].get("id") == req_id
```

Comment:
Seems to me that this should never happen. Instead, an error should be returned from `request_to_join_community`.

https://github.com/status-im/status-go/blob/cb3ff6bd607c9161ffab09a7715724565daa4b0a/protocol/messenger_communities.go#L1412-L1418

@osmaczko please correct me if I'm wrong.

Anyway, behaviour should be consistent, and the request should either be created or not. And tests should only expect 1 path.

Reply:
Error will be raised after https://github.com/status-im/status-go/pull/7178